### ISSUE TITLE
Why not asynchronously load WASM module?

### ISSUE THREAD

**Issue by cai-lw** [daboross/screeps-in-rust-via-wasm#9]
https://github.com/daboross/screeps-in-rust-via-wasm/blob/566af925b1af223fe5e6798cd31a90c5c93c96c8/cargo-screeps/src/build.rs#L235

Changing `false` to `true` cuts loading time (CPU cost in the tick when new codes are uploaded) from 100+ms down to ~1ms in the example AI, and I am not aware of any adverse effect.
I don't really know what's going on under the hood because I don't understand WASM at all. I would like to know why you use synchronous load here.

**Comment by daboross** [comment]
Generally because asynchronous things don't play well with Screeps' tick architecture. If it has the intended effect, we could switch to it. I had assumed that it would result in the tick ending as soon as the async part 'starts', then us not being able to run any intents on the first tick.

**Comment by Dessix** [comment]
This may cause attribution issues for CPU usage in the sandbox, as it allows "main" to return before you are actually done.

**Comment by cai-lw** [comment]
@Dessix I saw creeps mining and upgrading normally. I guess it somehow finished compiling and started working after a few ticks.
Spending 100+ms on compiling can be problematic. If the CPU bucket don't have enough time, the compilation will timeout and retry every tick, exhausting the bucket and getting itself stuck forever. It is not very common but I have met such scenario once.

**Comment by Dessix** [comment]
It's possible to check that the bucket is above a certain threshold before starting compilation, which is probably the correct move to make here. Async runs the same amount of work on the server, so if it's not charging you the same amount of CPU, it's a bug with the server implementation.

**Comment by daboross** [comment]
Yes- it will likely work *after* the initial tick, but doing an async load will still prevent it from running on that initial tick that it's loaded.

In async, it's still taking 100+ milliseconds, those 100+ milliseconds of processing are just happening after the server takes the final CPU count for your code. It's likely a bug that asynchronous processing in this way is allowed at all.

**Comment by cai-lw** [comment]
I think stopping for one tick is not a serious problem in most cases. There are other things that will case an unexpected stop, such as timeout hard reset, so generally codes should be prepared for it.
Ask Screeps and hope that it's not a bug.

**Comment by Dessix** [comment]
The fact that this doesn't charge you CPU for the compilation is a definitive sign that this is a bug.

I suggest:
- Reporting a minimal reproduction of the 100-vs-1 cpu sync/async case on the [Screeps server project](/screeps/screeps/issues)
- Adding a check to not begin WASM compilation when the [CPU bucket](http://docs.screeps.com/api/#Game.cpu) is below perhaps 9500.

**Comment by daboross** [comment]
We can definitely ask them @cai-lw. I believe this is likely a bug granting free CPU, but in the rare case that it isn't, we could use it.

Unfortunately I don't have much time this week. @Dessix if you wanted to, you could make an issue with the in the Screeps repo asking about this (and/or ask one of the community managers to look into it on slack).

Re: stopping WASM compilation. I'd rather not set such a high limit (9500)- I know many AIs have lower equilibrium they like to stay at, and I'd rather not stop compilation because of that. If you want to add a hardlimit in for Game.cpu.tickLimit<500, I'd be good with that- as long as we have a temp-`module.exports.loop` which will continue checking bucket amount and loading correctly once we have enough.

We could add a configuration for a higher limit than 500, but that's another issue altogether (user configuration of these things).

**Comment by cai-lw** [comment]
Sorry guys. I just found that I made a huge mistake that I put timing codes in Rust and it just doesn't work on the tick the module was asynchronously loaded.
When I move timing codes into JS, it doesn't show a significant difference between async load and sync load.

**Comment by daboross** [comment]
ah- it's alright. Thank you for investigating this thoroughly!

We probably still want to have some early-CPU cutoff so I'll leave this issue open until we have another one for that.

**Comment by ASalvail** [comment]
@daboross I believe this has been addressed with #142 ?

**Comment by daboross** [comment]
Sounds right to me! We deal with low-CPU situations fairly nicely now so this is good.

It might still be worth it to check if Screeps intends to leave the Promise-driven APIs available, but I guess that isn't _really_ an issue with this repository.
